nexus/src/app/sagas/instance_update/mod.rs

```diff
@@ -0,0 +1,439 @@
+// This Source Code Form is subject to the terms of the Mozilla Public
+// License, v. 2.0. If a copy of the MPL was not distributed with this
+// file, You can obtain one at https://mozilla.org/MPL/2.0/.
+
+use super::{
+    ActionRegistry, NexusActionContext, NexusSaga, SagaInitError,
+    ACTION_GENERATE_ID,
+};
+use crate::app::db::datastore::instance;
+use crate::app::db::datastore::InstanceSnapshot;
+use crate::app::db::lookup::LookupPath;
+use crate::app::db::model::Generation;
+use crate::app::db::model::VmmState;
+use crate::app::db::model::{Migration, MigrationState};
+use crate::app::sagas::declare_saga_actions;
+use chrono::Utc;
+use nexus_db_queries::{authn, authz};
+use omicron_common::api::external::Error;
+use omicron_uuid_kinds::GenericUuid;
+use omicron_uuid_kinds::InstanceUuid;
+use omicron_uuid_kinds::PropolisUuid;
+use serde::{Deserialize, Serialize};
+use steno::{ActionError, DagBuilder, Node};
+use uuid::Uuid;
+
+mod destroyed;
+use destroyed::*;
+
+// The public interface to this saga is actually a smaller saga that starts the
+// "real" update saga, which inherits the lock from the start saga. This is
+// because the decision of which subsaga(s) to run depends on the state of the
+// instance record read from the database *once the lock has been acquired*,
+// and the saga DAG for the "real" instance update saga may be constructed only
+// after the instance state has been fetched. However, since the the instance
+// state must be read inside the lock, that *also* needs to happen in a saga,
+// so that the lock is always dropped when unwinding. Thus, we have a second,
+// smaller saga which starts our real saga, and then the real saga, which
+// decides what DAG to build based on the instance fetched by the start saga.
+//
+// Don't worry, this won't be on the test.
+mod start;
+pub(crate) use self::start::{Params, SagaInstanceUpdate};
+
+/// Parameters to the "real" instance update saga.
+#[derive(Debug, Deserialize, Serialize)]
+struct RealParams {
+    serialized_authn: authn::saga::Serialized,
+
+    authz_instance: authz::Instance,
+
+    state: InstanceSnapshot,
+
+    orig_lock: instance::UpdaterLock,
+}
+
+const INSTANCE_LOCK_ID: &str = "saga_instance_lock_id";
+const INSTANCE_LOCK: &str = "updater_lock";
+const DESTROYED_VMM_ID: &str = "destroyed_vmm_id";
+const MIGRATION: &str = "migration";
```

Comment:
Yeah, I learnt this one painfully, as I'm sure you can imagine.

Unfortunately, the `declare_saga_actions!` doesn't currently allow a constant for the node name, so the string still has to be duplicated at least twice. I'd like to fix that at some point[^1], but it's unfortunately a bit more complex than just changing the macro to take an `$expr` instead of a `$literal` because we do some stuff with the `paste!` macro...

[^1]:  I _thought_ I'd opened an issue for it, but it looks like I forgot to...should probably go and do that.